proposals/0000-type-variable-scoping.rst

```diff
@@ -0,0 +1,1418 @@
+
+This component of this proposal revises the scoping of variables in GADT declaration syntax.
+
+Motivation
+^^^^^^^^^^
+
+1. This new version makes the distinction between universals and existentials in constructors
+   very clear, as required by the analysis `above <#universals-and-existentials>`_.
+
+#. Principled kind inference is, I believe, impossible using the current syntax. This is argued
+   in `Appendix B.8 <https://richarde.dev/papers/2020/kind-inference/kind-inference-supplement.pdf#subsection.B.8>`_ of the `Kind Inference for Datatypes`_ paper. The current approach allows some examples of polymorphic
+   recursion, but not others, and I doubt there is a declarative specification
+   of what we accept today. Here is an example of surprisingly allowed polymorphic
+   recursion::
+
+     data Poly a where
+       MkPoly :: forall k1 k2 (a :: k1) (b :: k2). Poly a -> Poly b -> Poly a
+
+   Note that the use of ``Poly b`` instantiates ``Poly`` at a different
+   kind (``k2``) than the instantiation in the result (``k1``).
+
+   If we change the ``Poly b`` above to ``Poly Maybe``, the definition is
+   rejected: only when the polymorphic recursion instantiates a kind variable
+   *with a variable* is the definition accepted.
+
+   Another oddity here happens when we ask what the inferred kind of ``Poly``
+   is, before generalization. It must be ``k1 -> Type``... but ``k1`` is not
+   even in scope in the declaration of ``Poly``. It's all very strange.
+
+Proposed Change Specification
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. Allow all variables -- call them ``a1 .. an`` -- in the header of a GADT declaration (including variables introduced implicitly) to scope over
+   all constructor declarations. Call the GADT ``G``.
+
+#. If there is no standalone kind signature for the GADT, each ``ai`` is
+   assigned a kind meta-variable. When checking the constructors, this kind
+   meta-variable is unified following the usual rules for meta-variables.
+   In particular, this meta-variable will not be able to unify with any kind
+   variable locally quantified in a constructor declaration, because the scope
+   of the locally quantified kind variable is smaller than the kind meta-variable.
+
+   If a variable kind signature (e.g. ``data G (a :: Type -> Type) where ...``)
+   is given in the GADT header, the kind given for the kind
+   variable is unified with the kind meta-variable, as usual.
+
+   In declarative terms, this means that we can simply "guess" a monokind for
+   each type variable ``ai``.
+
+#. For each constructor, we must determine the universal variables and the
+   existential variables. To do this, look at the result type ``G ty1 .. tyn``.
+   For each ``i`` such that ``tyi`` is exactly ``ai``, ``ai`` is labeled as
+   a universal variable for that constructor. All other type variables in
+   that constructor's type are existentials. The distinction between universals
+   and existentials matters only in patterns.
```

Comment:
This means that `MkT2` in 
```
data T2 a where MkT2 :: forall b -> T2 (Maybe b)
```
has an _existential_ type parameter, even though `b` seems to be no less of “input quality” than the `a` in
```
data T a where MkT :: forall a -> T a
```
Is that intended?

(I might be repeating arguments we already had when writing the paper back then. But I still think that trying to distinguish existentials and universals will cause many headaches like this.)

Reply:
The `b` in `MkT2` is definitely existential. Before matching on the pattern, all we know is that the scrutinee is of type `T2 ty` for some `ty`. It's the pattern-match that tells us that `ty` is `Maybe b` for some existent `b`.

Maybe you meant `MkT2 :: forall b -> T2 b`?

For the record, the arguments against the universal/existential distinction are growing, especially in the presence of visible dependent quantification. I just haven't tried to Think Hard about this yet.

Reply:
> It's the pattern-match that tells us that `ty` is `Maybe b` for some existent `b`.

Hmm, yes, but once we are in in this pattern, the value of `b` is fully determined by the value of `ty`. So in some sense it’s odd that seems odd to treat `b` as an “output of the match”.

> For the record, the arguments against the universal/existential distinction are growing, especially in the presence of visible dependent quantification. I just haven't tried to Think Hard about this yet.

Ok, will be patient then :-)

Reply:
(And will close this in favor of the more general thread https://github.com/ghc-proposals/ghc-proposals/pull/448#discussion_r739635748)